### PR TITLE
fix: secrets are properly validated with pattern in scaffolder template

### PR DESCRIPTION
**Pull request by Nikunj0601**
## Hey, I just made a Pull Request!

This PR solves the issue #23836. Secrets are currently validated with pattern by their masked value '***' instead of actual secret. So secrets are accessed in formData for validation and then deleted. 

Screenshot before:
image

after:
image

for both input is `testtest` and pattern: `^[a-z]{1}[a-z0-9]{2,7}$`

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx [hunk 1]
```diff
@@ -186,32 +189,31 @@ export const Stepper = (stepperProps: StepperProps) => {
...
-          analytics.captureEvent('click', `Next Step (${stepNum})`);
-          return stepNum;
-        });
+        setIsValidating(false);
+        return;
       }
+
+      const cleanedFormData = { ...formData };
+      Object.keys(secrets).forEach(secretKey => {
+        delete cleanedFormData[secretKey];
```
Comment (benjdlambert) [inline plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:205]: I think unfortunately, this only works for top level secret keys.

Imagine you have a schema like this:

```yaml
parameters:
  test:
    type: object
    required: [thing]
    properties:
      thing:
        ui:field: SecretField
```

Then this solution doesn't work right?

I think it's part of a bigger refactor of how we deal with the form state and secrets #28773

Comment (benjdlambert) [inline plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:205]: I think unfortunately, this only works for top level secret keys.

Imagine you have a schema like this:

```yaml
parameters:
  test:
    type: object
    required: [thing]
    properties:
      thing:
        ui:field: SecretField
```

Then this solution doesn't work right?

I think it's part of a bigger refactor of how we deal with the form state and secrets #28773 

I'd expect `${{ secrets.test.thing }}` to work here.

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Hey :wave: 

Thanks for trying to take a stab at this!

Bit of a problem I can see though at the minute that I think we wanna fix but it might require a lot of other work.

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx (hunk 1 above)
Comment (Nikunj0601) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:205]: I was looking into this with object schema type so it is working currently but not with `${{ secrets.test.thing }}` instead it is working for `${{ secrets.thing }}` cause the way secrets are currently stored is 
```
secrets: {
   thing: 'secretvalue',
}
```
```
formData: {
    "test": {
        "thing": "testtest"
    }
}
```

So, to make it work, I can think of two options: change the object storing to accommodate `${{ secrets.test.thing }}` or leave it as it is and in formData recursively check for secretKey to delete in formData after validation is completed.

I had one more question so when we click on create tasks api is called so for that payload do we need value of all the properties in formData even though one is secret?

Comment (Nikunj0601) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:205]: I was checking on recursively delete or mask secrets approach and it was working on the frontend side but it was throwing error from backend because in validate function only payloads property `values` but in which value of secret will either masked or deleted and it will throw an error. I was wondering should we need the value of secret field as it is or not?

Comment (benjdlambert) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:205]: Yeah, that's also the problem right, is that we need to split the schema apart into a secret schema to validate them individually. It all get a little bit messy and confusing, but maybe it's what we have to do unless we come up with a better way to define a secret schema or something?

Comment (Nikunj0601) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx]: I have pushed some changes in API to have a correct value in the secret field for the validation to be properly completed. Let me know if I have not understood it properly or if there are any changes to handle it in a better way.

Comment (Nikunj0601) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx]: @benjdlambert any feedback on the approach or it is not something you meant earlier.

Comment (Nikunj0601) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx]: @benjdlambert can you review the changes when you have a moment? I'd appreciate your feedback! Thanks!

Comment (benjdlambert) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx]: @Nikunj0601 sorry - haven't had chance to look at this yet - busy prepping for KubeCon. It's on my list though I haven't forgotten!

**Comment by Nikunj0601** [comment]
not stale

**Comment by Nikunj0601** [comment]
not stale

File: plugins/scaffolder-backend/src/service/router.ts [hunk 2]
```diff
@@ -604,7 +604,51 @@ export async function createRouter(
         );
 
         for (const parameters of [template.spec.parameters ?? []].flat()) {
-          const result = validate(values, parameters);
+          const validateWithSecrets = (
```
Comment (vinzscam) [inline plugins/scaffolder-backend/src/service/router.ts:607]: Thanks! Can you add some tests for this corner case? I'm not familiar with the shape of the data, so it will help me understand the problem 🙏

Perhaps we can also move these two new functions outside of the main route so that they are defined once.

File: plugins/scaffolder-backend/src/service/router.ts [hunk 3]
```diff
@@ -604,7 +604,51 @@ export async function createRouter(
...
+            parameterSchema: TemplateParametersV1beta3,
+            secrets: any,
+          ) => {
+            const valuesCopy = structuredClone(formValues);
+            console.log('parameters', parameterSchema);
+
+            const replaceMaskedSecrets = (
+              obj: any,
+              schema: any,
+              secretObj: any,
```
Comment (vinzscam) [inline plugins/scaffolder-backend/src/service/router.ts:618]: let's use proper typings here

**Review by vinzscam — CHANGES_REQUESTED** [review]

File: plugins/scaffolder-backend/src/service/router.ts (hunk 2 above)
Comment (Nikunj0601) [reply plugins/scaffolder-backend/src/service/router.ts]: I have added the test case and did moved those function outside. So currently when input is secret then it is masked value but actual value is sent in secrets field of payload. However while validating with condition it is using that masked value which was failing so i am replacing masked value with actual value from secrets field and then validating so it can poperly validate the input with conditions.

**Review by vinzscam — APPROVED** [review]
It looks good to me but I leave it to @backstage/scaffolder-maintainers for a final review 👍

**Comment by benjdlambert** [comment]
Need to have a think around this first before merging I think - just wanna cover some edge cases to make sure this works as expected :pray:

**Comment by Nikunj0601** [comment]
> Need to have a think around this first before merging I think - just wanna cover some edge cases to make sure this works as expected 🙏

Let me know if I have missed anything as I am not associated with any company which uses backstage so I might not know in detail about usecases of scaffolder templates.

**Comment by awanlin** [comment]
Hi @benjdlambert, just want to give you a nudge on this as the PR got closed but this seems like it's close to being able to ship.

**Comment by benjdlambert** [comment]
@awanlin @Nikunj0601 still haven't had time to get around to looking into this yet, also it's looking like #28773 is also becoming more relevant so not sure how this all fits together yet. Will try and get some time before next mainline to look at this. :pray: Apologies for the delay, we're focused on a lot of NFS stuff right now.

**Comment by Nikunj0601** [comment]
!not-closed
